src/storage/disk_manager.c

```diff
@@ -4410,28 +4419,38 @@ disk_reserve_from_cache (THREAD_ENTRY * thread_p, DISK_RESERVE_CONTEXT * context
       /* if we want to allocate temporary files, we have two options: preallocated permanent volumes (but with the
        * purpose of temporary files) or temporary volumes. try first the permanent volumes */
 
-      extend_info = &disk_Cache->temp_purpose_info.extend_info;
-
       if (disk_Cache->temp_purpose_info.nsect_perm_free > 0)
 	{
 	  disk_reserve_from_cache_vols (DB_PERMANENT_VOLTYPE, context);
-	}
-      if (context->n_cache_reserve_remaining <= 0)
-	{
-	  /* found enough sectors */
-	  assert (context->n_cache_reserve_remaining == 0);
-	  disk_cache_unlock_reserve_for_purpose (context->purpose);
-	  return NO_ERROR;
+	  if (context->n_cache_reserve_remaining <= 0)
+	    {
+	      /* found enough sectors */
+	      assert (context->n_cache_reserve_remaining == 0);
+	      disk_cache_unlock_reserve_for_purpose (context->purpose);
+	      return NO_ERROR;
+	    }
 	}
 
       /* reserve sectors from temporary volumes */
       extend_info = &disk_Cache->temp_purpose_info.extend_info;
-      if (extend_info->nsect_total - extend_info->nsect_free + context->n_cache_reserve_remaining
-	  >= disk_Temp_max_sects)
+      if (extend_info->nsect_free >= context->n_cache_reserve_remaining)
+	{
+	  disk_reserve_from_cache_vols (DB_TEMPORARY_VOLTYPE, context);
+	  if (context->n_cache_reserve_remaining <= 0)
+	    {
+	      /* found enough sectors */
+	      assert (context->n_cache_reserve_remaining == 0);
+	      disk_cache_unlock_reserve_for_purpose (context->purpose);
+	      return NO_ERROR;
+	    }
+	}
+
+      if (extend_info->nsect_total + context->n_cache_reserve_remaining >= disk_Temp_max_sects)
 	{
 	  /* too much temporary space */
 	  assert (false);
-	  er_set (ER_ERROR_SEVERITY, ARG_FILE_LINE, ER_BO_MAXTEMP_SPACE_HAS_BEEN_EXCEEDED, 1, disk_Temp_max_sects);
+	  er_set (ER_ERROR_SEVERITY, ARG_FILE_LINE, ER_BO_MAXTEMP_SPACE_HAS_BEEN_EXCEEDED, 1,
+		  (long) (disk_Temp_max_sects * DISK_SECTOR_NPAGES));
```

Comment:
still could be overflowed.
```suggestion
		  (long) disk_Temp_max_sects * DISK_SECTOR_NPAGES);
```

Reply:
thank your review, i will fix it.